SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -1601,3 +1601,43 @@ void USpatialNetDriver::HandleOnConnectionFailed(const FString& Reason)
 	UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Could not connect to SpatialOS. Reason: %s"), *Reason);
 	OnConnectionFailed.Broadcast(Reason);
 }
+
+uint32_t USpatialNetDriver::AddOpCallback(Worker_ComponentId ComponentId, const UserOpCallback& Callback)
+{
+	SpatialConstants::MIN_EXTERNAL_SCHEMA_ID <= ComponentId && ComponentId <= SpatialConstants::MAX_EXTERNAL_SCHEMA_ID;
+	uint32_t NewCallbackId = NextCallbackId++;
+	ComponentToCallbackIdMap.FindOrAdd(ComponentId).Add(NewCallbackId);
+	CallbackIdToDataMap.Add(NewCallbackId, UserOpCallbackData{ComponentId, Callback});
+	return NewCallbackId;
+}
+
+void USpatialNetDriver::RemoveOpCallback(uint32_t CallbackId)
+{
+	// Find callback ID in map and assert it does not exist
+	UserOpCallbackData CallbackDataToRemove = CallbackIdToDataMap.FindAndRemoveChecked(CallbackId);
+
+	TSet<uint32_t>* ComponentCallbackIds = ComponentToCallbackIdMap.Find(CallbackDataToRemove.ComponentId);
+	if (ComponentCallbackIds == nullptr)
+	{
+		UE_LOG(LogSpatialOSNetDriver, Warning, TEXT("Tried to remove callback ID: %d but found no callbacks for corresponding component ID: %d"), CallbackId, CallbackDataToRemove.ComponentId);
+	}
+
+	ComponentCallbackIds->Remove(CallbackId);
+}
+
+void USpatialNetDriver::RunUserCallbacks(Worker_ComponentId ComponentId, const Worker_Op* Op)
+{
+	if (TSet<uint32_t>* ComponentCallbackIds = ComponentToCallbackIdMap.Find(ComponentId))
```

Comment:
Our current style guide is to make nullptr conditionals explicit.

Reply:
I think TSet is a hash table. As calling callbacks is probably more likely than adding and removing them it might be worth sticking to the age old advice of "just use vector" for this.

Also I don't think we are expecting this set to be large.